### PR TITLE
Tests api processor routers

### PR DESCRIPTION
**Pull request by JFer11**
In this branch, first endpoints tests were added.
All "app" and "config" router endpoints were tested with pytest.

### CODE REVIEW

File: api/tests/app/test_app.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    return expected_response
+
+
+def expected_response_default():
+    """When no valid key is provided to the PUT request, config is updated with default values stored in AppDTO"""
+    expected_response = AppDTO().__dict__
+    expected_response = camel_case_to_snake_case_dict(expected_response)
+    expected_response = json_multi_type_to_json_string(expected_response)
+    expected_response["dashboard_url"] = expected_response["dashboard_u_r_l"]
```
Comment (pgrill) [inline api/tests/app/test_app.py:27]: I think that this line should be updated in the fuction `camel_case_to_snake_case_dict`.

File: api/tests/app/test_app.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    def test_get_app_config(self, config_rollback):
+        client, config_sample_path = config_rollback
+
+        response = client.get('/app')
+
+        expected_response = get_app_from_ini_config_file_json(config_sample_path)
+        expected_response["dashboardurl"] = expected_response["dashboard_url"]
+        del expected_response["dashboard_url"]
+
```
Comment (pgrill) [inline api/tests/app/test_app.py:44]: Why is this change required?

I tested the endpoint `/app` and it returns the key "dashboard_url"

File: api/tests/app/test_app.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        body = create_app_config(key_value_dict)
+
+        response = client.put('/app', json=body)
+
+        if correct_type == "string":
+            assert response.status_code == 200
+            expected_response = expected_response_function(config_sample_path)
+            assert response.json() == expected_response
+        elif correct_type == "integer":
```
Comment (pgrill) [inline api/tests/app/test_app.py:119]: Is missing the `correct_type == "integer"` case?

File: api/tests/app/test_app.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    # pytest -v api/tests/app/test_app.py::TestClassUpdateAppConfig::test_try_change_app_config_empty_json
+    def test_try_change_app_config_empty_json(self, config_rollback):
+        client, config_sample_path = config_rollback
+
+        body = {}
+        response = client.put('/app', json=body)
+
+        expected_response = expected_response_function(config_sample_path)
+        expected_response_2 = expected_response_default()
```
Comment (pgrill) [inline api/tests/app/test_app.py:158]: Maybe we can rename the variable as `default_response`?

File: api/tests/app/test_app.py (hunk 1 above)
Comment (JFer11) [reply api/tests/app/test_app.py:27]: I decided to include that line in the test instead of the function camel_case_to_snake_case_dict, to maintain consistency between the function name and its functionality. However, i can change it if you insist, just want you to know why i did it.

File: api/tests/app/test_app.py (hunk 2 above)
Comment (JFer11) [reply api/tests/app/test_app.py:44]: That is right, the endpoint returns he key "dashboard_url". However, my function get_app_from_ini_config_file_json, returns all the keys fine, less "dashboard_url", so we have to change it manually. Do not forget that i am changing "expected_response", not "response".

File: api/tests/app/test_app.py (hunk 1 above)
Comment (pgrill) [reply api/tests/app/test_app.py:27]: Mmmm, maybe I am wrong, but I think the snake_case of `dashboardURL` is `dashboard_url`. URL is an acronym, and in snake case we usually don't split them with `_` (for example API, csv, etc).
I think that is ok include this "border" case in the function camel_case_to_snake_case_dict.

File: api/tests/app/test_app.py (hunk 4 above)
Comment (JFer11) [reply api/tests/app/test_app.py:158]: Changed!

File: api/tests/app/test_app.py (hunk 3 above)
Comment (JFer11) [reply api/tests/app/test_app.py:119]: I think you meant the "bool" case, I made it explicit now.

**Review by pgrill — APPROVED** [review]

File: api/tests/app/test_app.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                    expected_response = get_app_from_config_file(config_sample_path)
+                    assert response.json() == expected_response
+        elif correct_type == "bool":
+            assert response.status_code == 400
+            assert response.json()['detail'][0]['type'] == 'type_error.' + "bool"
+        else:
+            assert response.status_code == 400
+
+    # pytest -v api/tests/app/test_app.py::TestClassUpdateAppConfig::test_try_change_app_config_non_existence_key
+    def test_try_change_app_config_non_existence_key(self, config_rollback):
```
Comment (renzodgc) [inline api/tests/app/test_app.py:121]: We could add a docstring or comment that extra keys are ignored

File: api/tests/app/test_config.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
+import pytest
+
+from api.tests.utils.common_functions import get_config_file_json, app_config_file_multi_type_json_to_string_json
+# The line below is absolutely necessary. Fixtures are passed as arguments to test functions. That is why IDE could
+# not recognized them.
```
Comment (renzodgc) [inline api/tests/app/test_config.py:5]:
```suggestion
# The line below is absolutely necessary. Fixtures are passed as arguments to test functions.
# This is why the IDE can't recognize them.
```

File: api/tests/app/test_config.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+from api.tests.utils.fixtures_tests import config_rollback, app_config
+
+
+def expected_response_update_report_info(config_sample_path):
+    app_config = get_config_file_json(config_sample_path)
+    expected_response = app_config_file_multi_type_json_to_string_json(app_config)
+
+    # It is weird that this change does not appear in response.json()
+    # Because if request was successfully, the PUT request will change the flag "has_been_configured"
```
Comment (renzodgc) [inline api/tests/app/test_config.py:14]: Should this be a `FIXME: ` comment?

File: api/tests/app/test_config.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    del body["app"]
+    return body
+
+
+# pytest -v api/tests/app/test_config.py::TestClassGetProcessorInfo
+class TestClassGetProcessorInfo:
+    """Get Config File, GET /config/info"""
+
+    def test_get_processor_info(self, config_rollback):
+        """I dont get why has been configured is True"""
```
Comment (renzodgc) [inline api/tests/app/test_config.py:31]: Should this also be a `FIXME` comment?

File: api/tests/utils/fixtures_tests.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import pytest
+import shutil
+import os
+
+from fastapi.testclient import TestClient
+
+# from api.tests.utils.common_variables import default
```
Comment (renzodgc) [inline api/tests/utils/fixtures_tests.py:7]: Should this be deleted?

File: api/tests/utils/fixtures_tests.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+def config_rollback():
+    original_path = '/repo/api/tests/data/config-x86-openvino.ini'
+    config_sample_path_to_modify = '/repo/api/tests/data/config-x86-openvino_TEMPORARY.ini'
+    shutil.copyfile(original_path, config_sample_path_to_modify)
+
+    config = ConfigEngine(config_sample_path_to_modify)
+    Settings(config=config)
+
+    # Here, to import ProcessorAPI successfully, "Settings" must be previously initialized with a config, that is why
+    # import order matters
```
Comment (renzodgc) [inline api/tests/utils/fixtures_tests.py:23]:
```suggestion
    # Import ProcessorAPI after Settings has been initialized with a config.
```

File: api/tests/app/test_config.py (hunk 8 above)
Comment (pgrill) [reply api/tests/app/test_config.py:31]: I think that it was fixed in the API refactor.

Comment (renzodgc) [reply api/tests/app/test_config.py:31]: So it's not an issue anymore?

Comment (pgrill) [reply api/tests/app/test_config.py:31]: Yes. we can remove the comment from the test

File: api/tests/README.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+###How tests were built
+
+To further improve the tests, I made this file to easily understand how they worked.
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/README.md:3]: Changed!

File: api/tests/README.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+The main thing to know is to understand what role fixtures play, especially Config_rollback.
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/README.md:5]: No, just Config_rollback

File: api/tests/README.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+###How tests were built
+
+To further improve the tests, I made this file to easily understand how they worked.
+
+The main thing to know is to understand what role fixtures play, especially Config_rollback. 
+
+Config_rollback fixture does the following:
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/README.md:7]: Changed!

File: api/tests/README.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+1. setUp: Initialize the app with a different config file (the one with template in the name).
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/README.md:8]: Changed!

File: api/tests/README.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+2. Returns the client to make the request and also returns the root of the config file. That root is also returned in order to go directly to that file and check if the changes made from our endpoints were valid.
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/README.md:9]: Thanks, changed!

File: api/tests/app/test_app.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
+import pytest
+
+from api.models.app import AppDTO
+from api.tests.utils.common_functions import create_app_config, camel_case_to_snake_case_dict, get_app_from_config_file
+# The line below is absolutely necessary. Fixtures are passed as arguments to test functions. That is why IDE could
+# not recognized them.
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/app/test_app.py:6]: Changed!

File: api/tests/app/test_app.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+        ({"weekly_global_report": "Here_must_be_a_bool_variable"}, "bool"),
+        ({"weekly_global_report": 40}, "bool"),
+        ({"heatmap_resolution": 40}, "string"),
+        ({"heatmap_resolution": False}, "string"),
+    ])
+    def test_try_change_app_config_wrong_type_variable_all(self, config_rollback, key_value_dict,
+                                                           correct_type):
+        """
+        It is important to mention, the fact that if pydantic expects a string, it will try to
+        cast what was sent, so if we send a bool or an integer to a string field, pydantic will work:
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/app/test_app.py:81]: Changed!

File: api/tests/app/test_app.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+          Example:
+              a = bool or integer
+              str(a)
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/app/test_app.py:84]: Changed!

File: api/tests/app/test_app.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+          Pydantic will accept it and PUT request will work.
+
+        Idem when we send a bool, and an integer is expected.
+            int(False) = 0
+            int(True) = 1
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/app/test_app.py:89]: Changed!

File: api/tests/app/test_app.py [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+                                                           correct_type):
+        """
+        It is important to mention, the fact that if pydantic expects a string, it will try to
+        cast what was sent, so if we send a bool or an integer to a string field, pydantic will work:
+          Example:
+              a = bool or integer
+              str(a)
+          Pydantic will accept it and PUT request will work.
+
+        Idem when we send a bool, and an integer is expected.
```
[the remark this answers is not in the document]
Comment (JFer11) [reply api/tests/app/test_app.py:87]: Changed!

File: api/tests/app/test_config.py (hunk 6 above)
Comment (JFer11) [reply api/tests/app/test_config.py:5]: Changed!

File: api/tests/app/test_app.py (hunk 5 above)
Comment (JFer11) [reply api/tests/app/test_app.py:121]: Changed!

File: api/tests/app/test_config.py (hunk 7 above)
Comment (JFer11) [reply api/tests/app/test_config.py:14]: This issue was reported and Pablo has already fixed it. :) The comment will be deleted.

File: api/tests/utils/fixtures_tests.py (hunk 10 above)
Comment (JFer11) [reply api/tests/utils/fixtures_tests.py:23]: Changed!

File: api/tests/utils/fixtures_tests.py (hunk 9 above)
Comment (JFer11) [reply api/tests/utils/fixtures_tests.py:7]: Yes, deleted!

**Review by renzodgc — APPROVED** [review]
